tokio/src/task/task_set.rs

```diff
@@ -0,0 +1,222 @@
+///     for i in 0..10 {
+///         assert!(seen[i]);
+///     }
+/// }
+/// ```
+pub struct TaskSet<T> {
+    inner: IdleNotifiedSet<JoinHandle<T>>,
+}
+
+impl<T> TaskSet<T> {
+    /// Create a new `TaskSet`.
+    pub fn new() -> Self {
+        Self {
+            inner: IdleNotifiedSet::new(),
+        }
+    }
+
+    /// Returns the number of tasks currently in the `TaskSet`.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns whether the `TaskSet` is empty.
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+}
+
+impl<T: 'static> TaskSet<T> {
+    /// Spawn the provided task on the task set.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if called outside of a Tokio runtime.
+    pub fn spawn<F>(&mut self, task: F)
+    where
+        F: Future<Output = T>,
+        F: Send + 'static,
+        T: Send,
+    {
+        self.insert(crate::spawn(task));
+    }
+
+    /// Spawn the provided task on the provided runtime and store it in this `TaskSet`.
+    pub fn spawn_on<F>(&mut self, task: F, handle: &Handle)
+    where
+        F: Future<Output = T>,
+        F: Send + 'static,
+        T: Send,
+    {
+        self.insert(handle.spawn(task));
+    }
+
+    /// Spawn the provided task on the current [`LocalSet`] and store it in this `TaskSet`.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if it is called outside of a `LocalSet`.
+    ///
+    /// [`LocalSet`]: crate::task::LocalSet
+    pub fn spawn_local<F>(&mut self, task: F)
+    where
+        F: Future<Output = T>,
+        F: 'static,
+    {
+        self.insert(crate::task::spawn_local(task));
+    }
+
+    /// Spawn the provided task on the provided [`LocalSet`] and store it in this `TaskSet`.
+    ///
+    /// [`LocalSet`]: crate::task::LocalSet
+    pub fn spawn_local_on<F>(&mut self, task: F, local_set: &LocalSet)
+    where
+        F: Future<Output = T>,
+        F: 'static,
+    {
+        self.insert(local_set.spawn_local(task));
+    }
+
+    fn insert(&mut self, jh: JoinHandle<T>) {
+        let mut entry = self.inner.insert_idle(jh);
+
+        // Set the waker that is notified when the task completes.
+        entry.with_value_and_context(|jh, ctx| jh.set_join_waker(ctx.waker()));
+    }
+
+    /// Wait until one of the tasks in the set completes and returns its output.
+    ///
+    /// Returns `None` if the set is empty.
+    ///
+    /// # Cancel Safety
+    ///
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `TaskSet`.
+    pub async fn join_one(&mut self) -> Result<Option<T>, JoinError> {
+        crate::future::poll_fn(|cx| self.poll_join_one(cx)).await
+    }
+
+    /// Abort all tasks on this `TaskSet`.
+    ///
+    /// This does not remove the tasks from the `TaskSet`. To wait for the tasks to complete
```

Comment:
I do feel that it would be more ergonomic for this to abort all tasks _and_ await there completion, but I would be alright with this design as well.

Reply:
We could have a shutdown function which first aborts all tasks and then returns a future which completes when all tasks have completed, and ignores their outputs, including ignoring any panics within the tasks.

Reply:
Well, right now, you would abort then call `join_one` until drained. There definitely should be a `!async` method for aborting. There could be one that both aborts & waits though.

Reply:
I would agree, it just might be nice to have this _and_ a more convenient method to wait for all to complete after the mass abort.